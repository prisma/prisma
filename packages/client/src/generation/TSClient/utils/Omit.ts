export function Omit(O: string, K: string, KName = 'P') {
  return `{ [${KName} in keyof (${O}) as ${KName} extends ${K} ? never : ${KName}]: (${O})[${KName}] }`
}
