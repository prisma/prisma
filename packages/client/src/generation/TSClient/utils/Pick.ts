export function Pick(O: string, K: string, KName = 'P') {
  return `{ [${KName} in keyof (${O}) as ${KName} extends ${K} ? ${KName} : never]: (${O})[${KName}] }`
}
