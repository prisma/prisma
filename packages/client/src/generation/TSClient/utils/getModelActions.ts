import { DMMF } from '@prisma/generator-helper'

import { DMMFHelper } from '../../../runtime/dmmf'

export function getModelActions(dmmf: DMMFHelper, name: string) {
  const mapping = dmmf.mappingsMap[name] ?? { model: name, plural: `${name}s` }

  const mappingKeys = Object.keys(mapping).filter(
    (key) => key !== 'model' && key !== 'plural' && mapping[key],
  ) as DMMF.ModelAction[]

  if ('aggregate' in mapping) {
    mappingKeys.push('count' as DMMF.ModelAction)
  }

  return mappingKeys
}
