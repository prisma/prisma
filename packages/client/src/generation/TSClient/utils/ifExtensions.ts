import { TSClient } from '../TSClient'

export function ifExtensions<R>(_case: R | (() => R), _default: R | (() => R)): R {
  if (TSClient.enabledPreviewFeatures.includes('clientExtensions')) {
    return typeof _case === 'function' ? (_case as any)() : _case
  }

  return typeof _default === 'function' ? (_default as any)() : _default
}
