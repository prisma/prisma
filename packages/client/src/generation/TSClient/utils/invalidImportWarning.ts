export const invalidImportWarningTs = [
  'You are importing your Prisma Client from the wrong path, please follow these steps:',
  '-----',
  '1. At the root of this project (or package), get the relative path to your client.',
  'For example, that could be `../prisma/client` or `./client` or something similar.',
  '-----',
  '2. Once you know where your client is located, add your own client as a dependency.',
  ' - For npm, run `npm add db@./path/to/your/client`',
  ' - For pnpm, run `pnpm add db@link:./path/to/your/client`',
  ' - For yarn, run `yarn add db@link:./path/to/your/client`',
  '-----',
  '3. If you have a `tsconfig.json`, you will need to set `module` and `moduleResolution`.',
  'Set `moduleResolution` to `nodenext` or `bundler`, and set `module` to `nodenext`.',
  '',
  '-----',
  '4. Update your imports, once done, your client will be properly set up.',
  '```',
  'import { PrismaClient } from "db"',
  '```',
  '-----',
  'More information: https://pris.ly/d/custom-output',
]

export const invalidImportWarningJs = invalidImportWarningTs.map((line) => line.replace('-----', '\\n'))
