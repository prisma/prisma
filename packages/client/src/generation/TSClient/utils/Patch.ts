import { Omit } from './Omit'

export function Patch(O1: string, O2: string, KName = 'P') {
  return `(${O1}) & ${Omit(O2, `keyof (${O1})`, KName)}`
}

export function Patch3(O1: string, O2: string, O3: string, KName = 'P') {
  return `(${O1}) & ${Omit(O2, `keyof (${O1})`, KName)} & ${Omit(O3, `keyof (${O1}) | keyof (${O2})`, KName)}`
}
