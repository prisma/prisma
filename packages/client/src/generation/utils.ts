import { assertNever } from '@prisma/internals'
import indent from 'indent-string'
import path from 'path'

import type { DMMFHelper } from '../runtime/dmmf'
import { DMMF } from '../runtime/dmmf-types'
import { ifExtensions } from './TSClient/utils/ifExtensions'

export enum Projection {
  select = 'select',
  include = 'include',
}

export function getScalarsName(modelName: string): string {
  return `${modelName}Scalars`
}

export function getPayloadName(modelName: string): string {
  return `${modelName}GetPayload`
}

// export function getExtractName(modelName: string, projection: Projection) {
//   return `Extract${modelName}${capitalize(projection)}`
// }

export function getSelectName(modelName: string): string {
  return `${modelName}Select`
}

export function getAggregateName(modelName: string): string {
  return `Aggregate${capitalize(modelName)}`
}

export function getGroupByName(modelName: string): string {
  return `${capitalize(modelName)}GroupByOutputType`
}

export function getAvgAggregateName(modelName: string): string {
  return `${capitalize(modelName)}AvgAggregateOutputType`
}

export function getSumAggregateName(modelName: string): string {
  return `${capitalize(modelName)}SumAggregateOutputType`
}

export function getMinAggregateName(modelName: string): string {
  return `${capitalize(modelName)}MinAggregateOutputType`
}

export function getMaxAggregateName(modelName: string): string {
  return `${capitalize(modelName)}MaxAggregateOutputType`
}

export function getCountAggregateInputName(modelName: string): string {
  return `${capitalize(modelName)}CountAggregateInputType`
}

export function getCountAggregateOutputName(modelName: string): string {
  return `${capitalize(modelName)}CountAggregateOutputType`
}

export function getAggregateInputType(aggregateOutputType: string): string {
  return aggregateOutputType.replace(/OutputType$/, 'InputType')
}

export function getGroupByArgsName(modelName: string): string {
  return `${capitalize(modelName)}GroupByArgs`
}

export function getGroupByPayloadName(modelName: string): string {
  return `Get${capitalize(modelName)}GroupByPayload`
}

export function getAggregateArgsName(modelName: string): string {
  return `${capitalize(modelName)}AggregateArgs`
}

export function getAggregateGetName(modelName: string): string {
  return `Get${capitalize(modelName)}AggregateType`
}

export function getAggregateScalarGetName(modelName: string): string {
  return `Get${capitalize(modelName)}AggregateScalarType`
}

export function getIncludeName(modelName: string): string {
  return `${modelName}Include`
}

export function getDefaultName(modelName: string): string {
  return `${modelName}Default`
}

export function getFieldArgName(field: DMMF.SchemaField, modelName: string): string {
  if (field.args.length) {
    return getModelFieldArgsName(field, modelName)
  }
  return getArgName((field.outputType.type as DMMF.OutputType).name)
}

export function getModelFieldArgsName(field: DMMF.SchemaField, modelName: string) {
  // Example: User$postsArgs
  // So it doesn't conflict with the generated type, like UserPostsArgs
  return `${modelName}$${field.name}Args`
}

export function getArgName(name: string): string {
  return `${name}Args`
}

// we need names for all top level args,
// as GraphQL doesn't have the concept of unnamed args
export function getModelArgName(modelName: string, action?: DMMF.ModelAction): string {
  if (!action) {
    return `${modelName}Args`
  }
  switch (action) {
    case DMMF.ModelAction.findMany:
      return `${modelName}FindManyArgs`
    case DMMF.ModelAction.findUnique:
      return `${modelName}FindUniqueArgs`
    case DMMF.ModelAction.findUniqueOrThrow:
      return `${modelName}FindUniqueOrThrowArgs`
    case DMMF.ModelAction.findFirst:
      return `${modelName}FindFirstArgs`
    case DMMF.ModelAction.findFirstOrThrow:
      return `${modelName}FindFirstOrThrowArgs`
    case DMMF.ModelAction.upsert:
      return `${modelName}UpsertArgs`
    case DMMF.ModelAction.update:
      return `${modelName}UpdateArgs`
    case DMMF.ModelAction.updateMany:
      return `${modelName}UpdateManyArgs`
    case DMMF.ModelAction.delete:
      return `${modelName}DeleteArgs`
    case DMMF.ModelAction.create:
      return `${modelName}CreateArgs`
    case DMMF.ModelAction.createMany:
      return `${modelName}CreateManyArgs`
    case DMMF.ModelAction.deleteMany:
      return `${modelName}DeleteManyArgs`
    case DMMF.ModelAction.groupBy:
      return `${modelName}GroupByArgs`
    case DMMF.ModelAction.aggregate:
      return getAggregateArgsName(modelName)
    case DMMF.ModelAction.count:
      return `${modelName}CountArgs`
    case DMMF.ModelAction.findRaw:
      return `${modelName}FindRawArgs`
    case DMMF.ModelAction.aggregateRaw:
      return `${modelName}AggregateRawArgs`
    default:
      assertNever(action, 'Unknown action')
  }
}

export function getFieldRefsTypeName(name: string) {
  return `${name}FieldRefs`
}

export function getDefaultArgName(dmmf: DMMFHelper, modelName: string, action: DMMF.ModelAction): string {
  const mapping = dmmf.mappings.modelOperations.find((m) => m.model === modelName)!

  const fieldName = mapping[action]
  const operation = getOperation(action)
  const queryType = operation === 'query' ? dmmf.queryType : dmmf.mutationType
  const field = queryType.fields.find((f) => f.name === fieldName)!
  return (field.args[0].inputTypes[0].type as DMMF.InputType).name
}

export function getOperation(action: DMMF.ModelAction): 'query' | 'mutation' {
  if (action === DMMF.ModelAction.findMany || action === DMMF.ModelAction.findUnique) {
    return 'query'
  }
  return 'mutation'
}

/**
 * Used to render the initial client args
 * @param modelName
 * @param fieldName
 * @param mapping
 */
export function renderInitialClientArgs( // TODO: dead code
  actionName: DMMF.ModelAction,
  fieldName: string,
  mapping: DMMF.ModelMapping,
): string {
  return `
  dmmf,
  fetcher,
  '${getOperation(actionName)}',
  '${fieldName}',
  '${mapping.plural}.${actionName}',
  args || {},
  [],
  errorFormat,
  measurePerformance\n`
}

export function getFieldTypeName(field: DMMF.SchemaField): string {
  if (typeof field.outputType.type === 'string') {
    return field.outputType.type
  }

  return field.outputType.type.name
}

export function getType(name: string, isList: boolean, isOptional?: boolean): string {
  return name + (isList ? '[]' : '') + (isOptional ? ' | null' : '')
}

export function getFieldType(field: DMMF.SchemaField): string {
  return getType(getFieldTypeName(field), field.outputType.isList)
}

interface SelectReturnTypeOptions {
  name: string
  actionName: DMMF.ModelAction
  renderPromise?: boolean
  hideCondition?: boolean
  isField?: boolean
  isChaining?: boolean
  fieldName?: string
  projection: Projection
}

/**
 * Get the complicated extract output
 * @param name Model name
 * @param actionName action name
 */
export function getReturnType({
  name,
  actionName,
  renderPromise = true,
  hideCondition = false,
  isField = false, // eslint-disable-line @typescript-eslint/no-unused-vars
  isChaining = false,
}: SelectReturnTypeOptions): string {
  if (actionName === 'count') {
    return `Promise<number>`
  }
  if (actionName === 'aggregate') return `Promise<${getAggregateGetName(name)}<T>>`

  if (actionName === 'findRaw' || actionName === 'aggregateRaw') {
    return `PrismaPromise<JsonObject>`
  }

  const isList = actionName === DMMF.ModelAction.findMany

  if (actionName === 'deleteMany' || actionName === 'updateMany' || actionName === 'createMany') {
    return `PrismaPromise<BatchPayload>`
  }

  /**
   * Important: We handle findMany or isList special, as we don't want chaining from there
   */
  if (isList || hideCondition) {
    const listOpen = isList ? 'Array<' : ''
    const listClose = isList ? '>' : ''
    const promiseOpen = renderPromise ? 'PrismaPromise<' : ''
    const promiseClose = renderPromise ? '>' : ''

    return `${promiseOpen}${listOpen}${getPayloadName(name)}<T${ifExtensions(', ExtArgs', '')}>${listClose}${
      isChaining ? '| Null' : ''
    }${promiseClose}`
  }

  if (actionName === 'findFirstOrThrow' || actionName === 'findUniqueOrThrow') {
    return `Prisma__${name}Client<${getType(
      getPayloadName(name) + `<T${ifExtensions(', ExtArgs', '')}>`,
      isList,
    )}${ifExtensions(', never, ExtArgs', '')}>`
  }
  if (actionName === 'findFirst' || actionName === 'findUnique') {
    if (isField) {
      return `Prisma__${name}Client<${getType(
        getPayloadName(name) + `<T${ifExtensions(', ExtArgs', '')}>`,
        isList,
      )} | Null${ifExtensions(', never, ExtArgs', '')}>`
    }
    return `HasReject<GlobalRejectSettings, LocalRejectSettings, '${actionName}', '${name}'> extends True ? Prisma__${name}Client<${getType(
      getPayloadName(name) + `<T${ifExtensions(', ExtArgs', '')}>`,
      isList,
    )}${ifExtensions(', never, ExtArgs', '')}> : Prisma__${name}Client<${getType(
      getPayloadName(name) + `<T${ifExtensions(', ExtArgs', '')}>`,
      isList,
    )} | null, null${ifExtensions(', ExtArgs', '')}>`
  }
  return `Prisma__${name}Client<${getType(
    getPayloadName(name) + `<T${ifExtensions(', ExtArgs', '')}>`,
    isList,
  )}${ifExtensions(', never, ExtArgs', '')}>`
}

export function isQueryAction(action: DMMF.ModelAction, operation: 'query' | 'mutation'): boolean {
  if (!(action in DMMF.ModelAction)) {
    return false
  }
  const result = action === DMMF.ModelAction.findUnique || action === DMMF.ModelAction.findMany
  return operation === 'query' ? result : !result
}

export function capitalize(str: string): string {
  return str[0].toUpperCase() + str.slice(1)
}

export function indentAllButFirstLine(str: string, indentation: number): string {
  const lines = str.split('\n')

  return lines[0] + '\n' + indent(lines.slice(1).join('\n'), indentation)
}

export function getRelativePathResolveStatement(outputDir: string, cwd?: string): string {
  if (!cwd) {
    return 'undefined'
  }
  return `path.resolve(__dirname, ${JSON.stringify(path.relative(outputDir, cwd))})`
}

/**
 * Returns unique elements of array
 * @param arr Array
 */

export function unique<T>(arr: T[]): T[] {
  const { length } = arr
  const result: T[] = []
  const seen = new Set() // just a cache

  loop: for (let i = 0; i < length; i++) {
    const value = arr[i]
    if (seen.has(value)) {
      continue loop
    }
    seen.add(value)
    result.push(value)
  }

  return result
}

export function getRefAllowedTypeName(type: DMMF.OutputTypeRef) {
  let typeName: string
  if (typeof type.type === 'string') {
    typeName = type.type
  } else {
    typeName = type.type.name
  }
  if (type.isList) {
    typeName += '[]'
  }

  return `'${typeName}'`
}
