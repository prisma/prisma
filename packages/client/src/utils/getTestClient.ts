import { getPlatform } from '@prisma/get-platform'
import {
  extractPreviewFeatures,
  getClientEngineType,
  getConfig,
  getEnvPaths,
  getRelativeSchemaPath,
  mapPreviewFeatures,
  parseEnvValue,
  printConfigWarnings,
} from '@prisma/internals'
import fs from 'fs'
import path from 'path'
import { parse } from 'stacktrace-parser'
import { promisify } from 'util'

import { getDMMF } from '../generation/getDMMF'
import type { GetPrismaClientConfig } from '../runtime/getPrismaClient'
import { getPrismaClient } from '../runtime/getPrismaClient'
import { ensureTestClientQueryEngine } from './ensureTestClientQueryEngine'
import { generateInFolder } from './generateInFolder'

const readFile = promisify(fs.readFile)

//TODO Rename to generateTestClientInMemory
/**
 * Returns an in-memory client for testing
 */
export async function getTestClient(schemaDir?: string, printWarnings?: boolean): Promise<any> {
  const callSite = path.dirname(require.main?.filename ?? '')
  const absSchemaDir = path.resolve(callSite, schemaDir ?? '')
  const schemaPath = await getRelativeSchemaPath(absSchemaDir)
  const datamodel = await readFile(schemaPath!, 'utf-8')
  const config = await getConfig({ datamodel, ignoreEnvVarErrors: true })
  if (printWarnings) {
    printConfigWarnings(config.warnings)
  }

  const generator = config.generators.find((g) => parseEnvValue(g.provider) === 'prisma-client-js')
  const previewFeatures = mapPreviewFeatures(extractPreviewFeatures(config))
  const platform = await getPlatform()
  const clientEngineType = getClientEngineType(generator!)

  await ensureTestClientQueryEngine(clientEngineType, platform)

  const document = await getDMMF({
    datamodel,
    previewFeatures,
  })
  const outputDir = absSchemaDir
  const relativeEnvPaths = getEnvPaths(schemaPath, { cwd: absSchemaDir })
  const activeProvider = config.datasources[0].activeProvider
  const options: GetPrismaClientConfig = {
    document,
    generator,
    dirname: absSchemaDir,
    relativePath: path.relative(outputDir, absSchemaDir),
    clientVersion: 'client-test-version',
    engineVersion: 'engine-test-version',
    relativeEnvPaths,
    datasourceNames: config.datasources.map((d) => d.name),
    activeProvider,
    dataProxy: Boolean(process.env.DATA_PROXY),
  }

  return getPrismaClient(options)
}

/**
 * Actually generates a test client with its own query-engine into ./@prisma/client
 */
export async function generateTestClient(projectDir?: string): Promise<any> {
  if (!projectDir) {
    const callsite = parse(new Error('').stack!)
    projectDir = path.dirname(callsite[1].file!)
  }

  await generateInFolder({
    projectDir,
    useLocalRuntime: false,
    transpile: true,
    useBuiltRuntime: false,
  })
}
