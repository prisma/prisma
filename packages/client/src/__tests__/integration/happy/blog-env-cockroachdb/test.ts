import sql from 'sql-template-tag'

import { generateTestClient } from '../../../../utils/getTestClient'
import type { SetupParams } from '../../../../utils/setupPostgres'
import { setupPostgres, tearDownPostgres } from '../../../../utils/setupPostgres'

const describeIf = (condition: boolean) => (condition ? describe : describe.skip)

describeIf(!process.env.TEST_SKIP_COCKROACHDB)('Blog fixture: Cockroachdb', () => {
  let prisma: any = null
  let PrismaHelpers: any = null
  let setupParams: any = null
  const requests: any[] = []
  const errorLogs: any[] = []

  beforeAll(async () => {
    await generateTestClient()
    const { PrismaClient, Prisma } = require('./node_modules/@prisma/client')
    PrismaHelpers = Prisma

    let originalConnectionString = process.env.TEST_COCKROACH_URI || 'postgresql://prisma@localhost:26257/tests'
    originalConnectionString += '-blog-env-cockroachdb'

    setupParams = {
      connectionString: originalConnectionString,
      dirname: __dirname,
    }

    await setupPostgres(setupParams)

    prisma = new PrismaClient({
      errorFormat: 'colorless',
      __internal: {
        measurePerformance: true,
        hooks: {
          beforeRequest: (request) => requests.push(request),
        },
      },
      datasources: {
        db: {
          url: originalConnectionString,
        },
      },
      log: [
        {
          emit: 'event',
          level: 'error',
        },
      ],
    })
  })

  afterAll(async () => {
    await prisma.$disconnect()
    await tearDownPostgres(setupParams.connectionString)
  })

  test('includes version in generated client', () => {
    const { Prisma } = require('./node_modules/@prisma/client')

    const { prismaVersion } = Prisma

    expect(prismaVersion).not.toBeUndefined()
    expect(prismaVersion.client).not.toBeUndefined()
  })

  test('does not leak connection strings in node_modules', () => {
    expect(prisma.internalDatasources).toBeUndefined()
  })

  test('can perform $queryRawUnsafe', async () => {
    const rawQuery = await prisma.$queryRawUnsafe('SELECT 1')
    expect(rawQuery[0]['?column?']).toBe(BigInt('1'))
  })

  test('can perform $queryRaw', async () => {
    const rawQuery = await prisma.$queryRaw`SELECT 1`
    expect(rawQuery[0]['?column?']).toBe(BigInt('1'))
  })

  test('Can do find query with client', async () => {
    const users1 = await prisma.user.findMany()
    expect(users1.length).toBe(1)
  })

  test('can throw validation errors', async () => {
    const {
      Prisma: { PrismaClientValidationError },
    } = require('./node_modules/@prisma/client')

    try {
      await prisma.post.create({
        data: {},
      })

      expect(false).toBe(true) // The line above needs to throw, so this should never be executed, but if it does (aka the line above did not throw, as expected), it will fail the test
    } catch (e) {
      expect(e).not.toBeUndefined()
      expect(e).toBeInstanceOf(PrismaClientValidationError)
    }
  })

  test('can run findMany queries', async () => {
    await prisma.post.create({
      data: {
        published: false,
        title: 'title',
        content: 'content',
      },
    })
    const posts = await prisma.post.findMany()
    expect(posts).not.toHaveLength(0)
  })

  test('can run findMany queries with a `null` where', async () => {
    await prisma.post.create({
      data: {
        published: false,
        title: 'null where',
        content: null,
      },
    })
    const posts = await prisma.post.findMany({
      where: {
        content: null,
      },
    })
    expect(posts.length).not.toBe(0)
  })

  test('can run create queries', async () => {
    const post = await prisma.post.create({
      data: {
        published: false,
        title: 'Some title',
      },
    })

    expect(post).not.toBeUndefined()
  })

  test('can run delete queries', async () => {
    const post = await prisma.post.create({
      data: {
        published: false,
        title: 'Some title',
      },
    })
    const deletedPost = await prisma.post.delete({
      where: { id: post.id },
      select: {
        authorId: true,
        content: true,
        published: true,
        title: true,
      },
    })

    expect(deletedPost).toMatchInlineSnapshot(`
      {
        authorId: null,
        content: null,
        published: false,
        title: Some title,
      }
    `)
  })

  test('can run update queries', async () => {
    const post = await prisma.post.create({
      data: {
        published: false,
        title: 'Some title',
      },
    })
    const updatedPost = await prisma.post.update({
      where: {
        id: post.id,
      },
      data: {
        title: 'Updated title',
      },
      select: {
        authorId: true,
        content: true,
        published: true,
        title: true,
      },
    })

    expect(updatedPost).toMatchInlineSnapshot(`
      {
        authorId: null,
        content: null,
        published: false,
        title: Updated title,
      }
    `)

    const resultJsonArray = await prisma.post.create({
      data: {
        published: false,
        title: 'json array',
        jsonData: [
          {
            array1key: 'array1value',
          },
        ],
      },
    })
    const result = await prisma.post.findMany({
      where: {
        jsonData: {
          equals: [
            {
              array1key: 'array1value',
            },
          ],
        },
      },
    })
    expect(result.length).toBe(1)

    const resultJsonUpdateWithSet = await prisma.post.update({
      where: {
        id: resultJsonArray.id,
      },
      data: {
        title: 'json array updated 2',
        jsonData: {
          set: [
            {
              array1key: 'array1valueupdated',
            },
          ],
        },
        coinflips: {
          set: [true, true, true, false, true],
        },
      },
      select: {
        authorId: true,
        coinflips: true,
        content: true,
        jsonData: true,
        published: true,
        title: true,
      },
    })
    expect(resultJsonUpdateWithSet).toMatchInlineSnapshot(`
      {
        authorId: null,
        coinflips: [
          true,
          true,
          true,
          false,
          true,
        ],
        content: null,
        jsonData: {
          set: [
            {
              array1key: array1valueupdated,
            },
          ],
        },
        published: false,
        title: json array updated 2,
      }
    `)
  })

  describe('$queryRaw', () => {
    test('$queryRaw(sql`<SQL>`)', async () => {
      await prisma.user.create({ data: { email: 'c@a.de', name: 'C' } })
      const users = await prisma.$queryRaw(sql`SELECT * FROM "public"."User"`)
      expect(users).not.toHaveLength(0)
    })

    test('$queryRaw(sql`<SQL>`) with params', async () => {
      await prisma.user.create({ data: { email: 'd@a.de', name: 'D' } })
      const users = await prisma.$queryRaw(sql`SELECT * FROM "public"."User" WHERE name = ${'D'}`)
      expect(users).not.toHaveLength(0)
    })

    test('$queryRaw`<SQL>`', async () => {
      await prisma.user.create({ data: { email: 'e@a.de', name: 'E' } })
      const users = await prisma.$queryRaw`SELECT * FROM "public"."User"`
      expect(users).not.toHaveLength(0)
    })

    test('$queryRaw`<SQL>` with join', async () => {
      const users = await prisma.$queryRaw`SELECT * FROM "public"."User" WHERE id IN (${PrismaHelpers.join([
        '42',
        '333',
        '2048',
      ])})`
      expect(users).toHaveLength(0)
    })

    test('$queryRaw`<SQL>` with params', async () => {
      await prisma.user.create({ data: { email: 'f@a.de', name: 'F' } })
      const users = await prisma.$queryRaw`SELECT * FROM "public"."User" WHERE name = ${'F'}`
      expect(users[0].name).toBe('F')
    })

    test('$queryRaw(string) error', async () => {
      const users = prisma.$queryRaw('<strings will throw>')

      await expect(users).rejects.toThrowErrorMatchingSnapshot()
    })
  })

  test('$executeRaw', async () => {
    // Test executeRaw(string)
    const rawexecute = await prisma.$executeRawUnsafe('SELECT 1')
    expect(rawexecute).toBe(1)

    // Test executeRaw``
    const rawexecuteTemplate = await prisma.$executeRaw`SELECT 1`
    expect(rawexecuteTemplate).toBe(1)

    // Test executeRaw`` with ${param}
    const rawexecuteTemplateWithParams = await prisma.$executeRaw`SELECT * FROM "public"."User" WHERE name = ${'Alice'}`
    expect(rawexecuteTemplateWithParams).toBe(1)
  })
})
