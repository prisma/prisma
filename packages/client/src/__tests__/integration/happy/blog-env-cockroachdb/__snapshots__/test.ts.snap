// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blog fixture: Cockroachdb $queryRaw $queryRaw(string) error 1`] = `
\`$queryRaw\` is a tag function, please use it like the following:
\`\`\`
const result = await prisma.$queryRaw\`SELECT * FROM User WHERE id = \${1} OR email = \${'user@email.com'};\`
\`\`\`

Or read our docs at https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access#queryraw

`;
